lib/elixir/lib/calendar.ex

```diff
@@ -363,4 +363,504 @@ defmodule Calendar do
+  # Year
+  defp format_modifiers("Y" <> rest, width, pad, datetime, format_options, acc) do
+    result = datetime.year |> Integer.to_string() |> pad_leading(width, pad)
+    parse(rest, datetime, format_options, [result | acc])
+  end
+
+  # +hhmm/-hhmm time zone offset from UTC (empty string if naive)
+  defp format_modifiers(
+         "z" <> rest,
+         width,
+         pad,
+         datetime = %{utc_offset: utc_offset, std_offset: std_offset},
+         format_options,
+         acc
+       ) do
+    absolute_offset = abs(utc_offset + std_offset)
+
+    offset_number =
+      Integer.to_string(div(absolute_offset, 3600) * 100 + rem(div(absolute_offset, 60), 60))
+
+    sign = if utc_offset + std_offset >= 0, do: "+", else: "-"
+    result = "#{sign}#{pad_leading(offset_number, width, pad)}"
+    parse(rest, datetime, format_options, [result | acc])
+  end
+
+  defp format_modifiers("z" <> rest, _width, _pad, datetime, format_options, acc) do
+    parse(rest, datetime, format_options, ["" | acc])
+  end
+
+  # Time zone abbreviation (empty string if naive)
+  defp format_modifiers("Z" <> rest, width, pad, datetime, format_options, acc) do
+    result = datetime |> Map.get(:zone_abbr, "") |> pad_leading(width, pad)
+    parse(rest, datetime, format_options, [result | acc])
+  end
+
+  defp pad_preferred(result, width, pad) when length(result) < width do
+    pad_preferred([pad | result], width, pad)
+  end
+
+  defp pad_preferred(result, _width, _pad), do: result
+
+  defp pad_leading(string, count, padding) do
+    to_pad = count - byte_size(string)
+    if to_pad > 0, do: do_pad_leading(to_pad, padding, string), else: string
+  end
+
+  defp do_pad_leading(0, _, acc), do: acc
+
+  defp do_pad_leading(count, padding, acc),
+    do: do_pad_leading(count - 1, padding, [padding | acc])
+
+  defp options(user_options) do
+    default_options = %{
+      preferred_date: "%Y-%m-%d",
+      preferred_time: "%H:%M:%S",
+      preferred_datetime: "%Y-%m-%d %H:%M:%S",
+      am_pm_names: fn
+        :am -> "am"
+        :pm -> "pm"
+      end,
+      month_names: fn month ->
+        {"January", "February", "March", "April", "May", "June", "July", "August", "September",
+         "October", "November", "December"}
+        |> elem(month - 1)
+      end,
+      day_of_week_names: fn day_of_week ->
+        {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"}
+        |> elem(day_of_week - 1)
+      end,
+      abbreviated_month_names: fn month ->
+        {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}
+        |> elem(month - 1)
+      end,
+      abbreviated_day_of_week_names: fn day_of_week ->
+        {"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"} |> elem(day_of_week - 1)
+      end,
+      preferred_datetime_invoked: false,
+      preferred_date_invoked: false,
+      preferred_time_invoked: false
+    }
+
+    Enum.reduce(user_options, default_options, fn {key, value}, acc ->
```

Comment:
I think this could be changed to a `Map.merge`?

Reply:
It can be `Enum.into(user_options, default_options)`, although it would add mismatched keys to the final map. Maybe it makes more sense to keep it a reduce and raise in the else branch instead. Thoughts?